File: Longest.java
// Time Complexity : O(N^2)
// Space Complexity : O(N)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No
class Solution {
    public int lengthOfLIS(int[] nums) {
        if(nums == null || nums.length == 0){
            return 0;
        }
        
        int max  = 1;
        int[] dp = new int[nums.length];
        Arrays.fill(dp,1);
        
        for(int i = 1; i < dp.length; i++){
            for(int j = 0; j < i; j++){
                if(nums[i] > nums[j]){
                    dp[i] = Math.max(dp[i], dp[j] + 1);
                    max = Math.max(max,dp[i]);
                }
            }
        }
        
        return max;
    }
}

File: longest.py
# Time: O(n^2)
# Space: O(n) 
# Did it run on Leetcode: yes

class Solution(object):
    def lengthOfLIS(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        # n=len(nums)
        # dp=[1]*(n)
        # maxel=1
        # for i in range(1,n):
        #     for j in range(i):
        #         if(nums[i]>nums[j]):
        #             dp[i]=max(dp[i],dp[j]+1)
        #             maxel=max(maxel, dp[i])
        # return maxel

        n=len(nums)
        dp=[1]*(n)
        maxel=1
        for i in range(1,n):
            for j in range(i):
                if(nums[i]>nums[j]):
                    # allsubs.append(list(nums[j]))
                    dp[i]=max(dp[i],dp[j]+1)
                    maxel=max(maxel, dp[i])
        # print(allsubs)
        return maxel